File: python/oz/ozhtml/ast.py
from oz.ozhtml.ast_const import AstConst


class Ast:
    def __init__(self, ast_index: int, value):
        self.ast_index = ast_index
        self.value = value
        self.name = AstConst.NAMES[ast_index]

    def _is(self, key, value):
        if isinstance(key, list):
            b = None
            for iii in key:
                b = self._is(self, iii, value)
                if b:
                    b = True
                    break
            return b
        return self.astIndex == key

    def is_not(self, key, value):
        return self._is(key, value) is None

    def expected(self, key, value):
        b = self._is(key, value)
        if b is None:
            raise Exception("unexpected ast " + self.name);
        return b

    def unexpected(self, key, value):
        b = self.is_not(key, value)
        if b is None:
            raise Exception("unexpected ast " + self.name);
        return b

File: go/oz/ozhtml/ast.go
package ozhtml

import (
	"errors"
	"oz\ast_const"
)

type Ast struct {
	astIndex int
	value    string
	name     string
}



func is(ast *Ast, astIndexs [] int, value string) bool {
	for astIndex := range astIndexs {
		b1 := ast.astIndex==astIndex
		b2 := ast.value== value
		if b1 && b2 {
			return  true
		}
	}
	return  false
}
func isNot(ast *Ast, astIndexs[] int, value string) bool {
	return !is(ast, astIndexs, value)
}
func expected(ast *Ast, astIndexs[] int, value string) (r bool, err error) {
	b := is(ast, astIndexs, value)
	if b {
		return b, errors.New("unexpected ast " + ast.name)
	}
	return b, nil
}
func unexpected(ast *Ast,  astIndexs[] int, value string) (r bool, err error) {
	b := isNot(ast, astIndexs, value)
	if b {
		return b, errors.New("unexpected ast " + ast.name)
	}
	return b, nil
}
func toString(ast *Ast) string {
	if ast.value == "" {
		return "Ast." + ast.name + "();";
	} else {
		return "Ast." + ast.name + "();";
	}
}